fendermint/vm/topdown/src/finality.rs

```diff
@@ -0,0 +1,426 @@
+#[derive(Clone)]
+struct ParentViewData {
+    height_data: TVar<SequentialKeyCache<BlockHeight, (Bytes, ValidatorSet)>>,
+    top_down_msgs: TVar<SequentialKeyCache<Nonce, CrossMsg>>,
+}
+
+impl ParentViewData {
+    fn latest_height(&self) -> StmResult<Option<BlockHeight>> {
+        let cache = self.height_data.read()?;
+        // safe to unwrap, we dont allow no upper bound
+        Ok(cache.upper_bound())
+    }
+
+    fn block_hash(&self, height: BlockHeight) -> StmResult<Option<Bytes>> {
+        let cache = self.height_data.read()?;
+        Ok(cache.get_value(height).map(|i| i.0.clone()))
+    }
+
+    fn validator_set(&self, height: BlockHeight) -> StmResult<Option<ValidatorSet>> {
+        let cache = self.height_data.read()?;
+        Ok(cache.get_value(height).map(|i| i.1.clone()))
+    }
+
+    fn all_top_down_msgs(&self) -> StmResult<Vec<CrossMsg>> {
+        let cache = self.top_down_msgs.read()?;
+        Ok(cache.values().cloned().collect())
+    }
+}
+
+// TODO: keep it first, might be useful later
+// macro_rules! downcast_err {
+//     ($r:ident) => {
+//         match $r {
+//             Ok(()) => Ok(()),
+//             Err(e) => match e.downcast_ref::<Error>() {
+//                 None => unreachable!(),
+//                 Some(e) => Err(e.clone())
+//             }
+//         }
+//     }
+// }
+
+#[async_trait]
+impl ParentViewProvider for DefaultFinalityProvider {
+    async fn latest_height(&self) -> StmDynResult<Option<BlockHeight>> {
+        let h = self.parent_view_data.latest_height()?;
+        Ok(h)
+    }
+
+    async fn latest_nonce(&self) -> StmDynResult<Option<Nonce>> {
+        let top_down_msgs = self.parent_view_data.top_down_msgs.read()?;
+        Ok(top_down_msgs.upper_bound())
+    }
+
+    async fn new_block_height(
+        &self,
+        height: BlockHeight,
+        block_hash: BlockHash,
+        validator_set: ValidatorSet,
+    ) -> StmDynResult<()> {
+        let insert_res = self.parent_view_data.height_data.modify(|mut cache| {
+            let r = cache.append(height, (block_hash.clone(), validator_set.clone()));
+            (cache, r)
+        })?;
+
+        match insert_res {
+            Ok(_) => Ok(()),
+            Err(_) => {
+                // now the inserted height is not the next expected block height, could be a chain
+                // reorg if the caller is behaving correctly.
+                abort(Error::ParentReorgDetected(height))
+            }
+        }
+    }
+
+    async fn new_top_down_msgs(&self, top_down_msgs: Vec<CrossMsg>) -> StmDynResult<()> {
+        if top_down_msgs.is_empty() {
+            // not processing if there are no top down msgs
+            return Ok(());
+        }
+
+        // get the min nonce from the list of top down msgs and purge all the msgs with nonce
+        // about the min nonce in cache, as the data should be newer and more accurate.
+        let min_nonce = top_down_msgs.first().unwrap().msg.nonce;
```

Comment:
This is handling nonce overlaps. If there are nonce overlaps, we take the new version. We do this by deleting all previous overlapped top down messages and insert new ones. Currently we cannot update as part of block height because top down messages are stored by nonce, the key of the solidity mapping is nonce, not by block height.

Reply:
This is completely updated with new top down messages query, this mechanism is not present anymore.